### PR TITLE
Enables setting optimization objectives from ompl_planning.yaml

### PR DESCRIPTION
**Pull request by rbbg**
Not sure if this is the best way to do it, but I have seen it requested a number of times and I think it's a fairly minor change.

### CODE REVIEW

**Comment by v4hn** [comment]
Is this based on an option already supported in ompl_planning.yaml?
If it is not, I would really prefer to have the actual names of the objectives in the config instead of a magic number.

**Comment by rbbg** [comment]
Yes that does seem better. Thanks.

One other question: It seems to me like the default PathLengthOptimization is optimizing the 1-norm distance in jointspace when used by MoveIt!. Is this by choice? The 1-norm seems to be a little odd as a default optimization objective, at least from my point of view.

File: ompl/ompl_interface/src/model_based_planning_context.cpp [hunk 1]
```diff
@@ -217,6 +224,38 @@ void ompl_interface::ModelBasedPlanningContext::useConfig()
...
     return;
 
+
+  ompl::base::OptimizationObjectivePtr objective;
+  it = cfg.find("optimization_objective");
+  if (it == cfg.end())
+  {
+    logInform("%s: No optimization objective specified, defaulting to PathLengthOptimizationObjective");
+    objective.reset(new ompl::base::PathLengthOptimizationObjective(ompl_simple_setup_->getSpaceInformation()));
+  }
```
Comment (v4hn) [inline ompl/ompl_interface/src/model_based_planning_context.cpp:244]: could you rewrite the default case to something like this:

```
std::string objective;
it = cfg.find("optimization_objective");
if (it == cfg.end())
{
   objective = "PathLengthOptimizationObjective";
   logInform(" ... defaulting to " + objective); // with appropriate string type conversions I suppose...
}
else
{
  objective = it->second;
}
```

This avoids redundancy and the big else-block afterwards.

**Comment by v4hn** [comment]
> It seems to me like the default PathLengthOptimization is optimizing the 1-norm distance in jointspace when used by MoveIt!. Is this by choice?

I don't know @davetcoleman is our OMPL expert afaik :-)
But what norm would you prefer in joint space?

**Comment by rbbg** [comment]
Well, not sure what that was based on, but I always kind of assumed that it was using the 2-norm to begin with. I don't think the 2-norm is _better_, but I've talked to at least a couple of other people that had made the same assumption, hence the question.

**Comment by davetcoleman** [comment]
Yea, I was surprised recently by the discovery it was using 1-norm also - it really threw off some assumptions I was making in my research! I then assumed that in joint space, 1-norm is the only distance metric that makes sense? I tried to find some publications to back this up, but was unsuccessful. @isucan @mamoll - why does the MoveIt! [distance function](https://github.com/ros-planning/moveit_core/blob/jade-devel/robot_model/src/revolute_joint_model.cpp#L161) used in OMPL only use the 1-norm?

**Comment by davetcoleman** [comment]
Back to this PR: thanks for creating this PR! Could you provide some documentation for users who want to use this exposed configuration? It would be best if you added a folder to this repo: `moveit_planners/ompl/ompl_interface/tutorials/doc/tutorial.rst` and we could then link to it from [http://moveit.ros.org/documentation/tutorials/]. Its ok if the tutorial is really short for now, it would be a good start. For reference see [here](https://github.com/ros-planning/moveit_ros/blob/jade-devel/visualization/doc/tutorial.rst)

**Comment by mamoll** [comment]
@davetcoleman, the distance function you link to computes the distance for a single joint. The real "problem" arises when you try to combine distances from individual joints. The current approach is (most likely) to simply add the distances over individual joints together. This is simple and computationally efficient. More important than 1-norm vs. 2-norm is probably how the individual joint distances are weighted. A shoulder displacement typically has a much larger effect than bending a finger. Is this accounted for in MoveIt!?

It is not obvious to me what the "right" distance measure should be articulated mechanisms (or whether 2-norm would somehow be better than 1-norm). Swept volume seems like a reasonable choice, but tends to be expensive to compute. In my experience, motion planning algorithms that depend heavily on a distance metric tend to scale badly with increasing number of degrees of freedom.

**Comment by davetcoleman** [comment]
@mamoll you are right that the [current approach](https://github.com/ros-planning/moveit_core/blob/jade-devel/robot_model/src/joint_model_group.cpp#L409) is to simply "add the distances over the individual joints together". @isucan added a notion of `distance_factor_` but for all joint types it is always set to just `1`. 

> A shoulder displacement typically has a much larger effect than bending a finger. Is this accounted for in MoveIt!?

In MoveIt! all joints are weighted the same, there there is a related notion of a `projection_evaluator` inside `moveit_planners_ompl` that takes i.e. the shoulder join's translation and uses it in some planners for better distance functions (I think).

Thanks for all that insight @mamoll, very interesting!

**Comment by davetcoleman** [comment]
This tutorial looks great, thanks! Travis failed due to a rosdep error - I just retriggered it. When it passes +1

**Comment by davetcoleman** [comment]
We need to add a link to this tutorial [here](https://github.com/julienking/julienking.github.io/blob/master/documentation/tutorials/index.markdown) once rosdoc_lite generates the new webpage (takes about 24 hours)

**Comment by davetcoleman** [comment]
We forgot the additional steps to turn on rosdoc, added here: https://github.com/ros-planning/moveit_planners/issues/76

**Comment by davetcoleman** [comment]
Cherry-picked to jade & kinetic

**Comment by davetcoleman** [comment]
I've added a link to the documentation here https://github.com/julienking/julienking.github.io/pull/19

**Pull request merged**